src/coroutine/waker.rs

```diff
@@ -0,0 +1,96 @@
+use crate::sync::GILOnceCell;
+use crate::types::PyCFunction;
+use crate::{intern, wrap_pyfunction, Py, PyAny, PyObject, PyResult, Python};
+use futures_task::ArcWake;
+use pyo3_macros::pyfunction;
+use std::sync::Arc;
+
+/// Lazy `asyncio.Future` wrapper, implementing [`ArcWake`] by calling `Future.set_result`.
+///
+/// asyncio future is let uninitialized until [`initialize_future`][1] is called.
+/// If [`wake`][2] is called before future initialization (during Rust future polling),
+/// [`initialize_future`][1] will return `None` (it is roughly equivalent to `asyncio.sleep(0)`)
+///
+/// [1]: AsyncioWaker::initialize_future
+/// [2]: AsyncioWaker::wake
+pub struct AsyncioWaker(GILOnceCell<Option<LoopAndFuture>>);
+
+impl AsyncioWaker {
+    pub(super) fn new() -> Self {
+        Self(GILOnceCell::new())
+    }
+
+    pub(super) fn reset(&mut self) {
+        self.0.take();
+    }
+
+    pub(super) fn initialize_future<'a>(&'a self, py: Python<'a>) -> PyResult<Option<&'a PyAny>> {
+        let init = || LoopAndFuture::new(py).map(Some);
+        let loop_and_future = self.0.get_or_try_init(py, init)?.as_ref();
+        Ok(loop_and_future.map(|LoopAndFuture { future, .. }| future.as_ref(py)))
+    }
+}
+
+impl ArcWake for AsyncioWaker {
+    fn wake_by_ref(arc_self: &Arc<Self>) {
+        Python::with_gil(|gil| {
+            if let Some(loop_and_future) = arc_self.0.get_or_init(gil, || None) {
+                loop_and_future
+                    .set_result(gil)
+                    .expect("unexpected error in coroutine waker");
+            }
+        });
+    }
+}
+
+struct LoopAndFuture {
+    event_loop: PyObject,
+    future: PyObject,
+}
+
+impl LoopAndFuture {
+    fn new(py: Python<'_>) -> PyResult<Self> {
+        static GET_RUNNING_LOOP: GILOnceCell<PyObject> = GILOnceCell::new();
```

Comment:
I don't think caching the running loop and storing within each future is useful. If we assume that loop never changes, we can just access `GET_RUNNING_LOOP` from e.g. within `LoopAndFuture::set_result` instead of storing the loop within each instance.

Additionally I would suggest using a single API cache, e.g.

```rust
struct AsyncioApi {
  create_future: PyObject,
  call_soon_threadsafe: PyObject,
  is_closed: PyObject,
  release_waiter: PyCFunction,
}

impl AsyncioApi {
  fn init(py: Python) -> PyResult<Self> {
    let asyncio = py.import("asyncio")?;
    let loop = asyncio.getattr("get_running_loop")?.into();
    let create_future = loop.getattr("create_future")?.into();
    let call_soon_threadsafe = loop.getattr("call_soon_threadsafe")?.into();
    let is_closed = loop.getattr("is_closed")?.into();
    let release_waiter = wrap_pyfunction!(release_waiter, py);

    Ok(Self { .. })
  }
}

static ASYNCIO_API: GILOnceCell<AsyncioApi> = GILOnceCell::new();

impl AsyncioFuture {
  fn new(py: Python) -> PyResult<Self> {
     let api = ASYNCIO_API.get_or_try_init(py, AsyncioApi::init)?;
     let future = api.create_future.call0(py)?; 

     Ok(Self { future })
  }
}
```

Reply:
I'm not convinced we should make the assumption the loop never changes, at least there will be one event loop per thread and I'm sure there must be software out there which has multiple event loops per thread for some arcane reason.

Reply:
I don't assume the loop will never change, I assume it can change (thanks to you @adamreichold by the way). That's why I don't think we can use a cache with event loop method. Also, you can have different event loop, one thread with asyncio, another with uvloop, everything is possible.

`LoopAndFuture::set_result` is called in `Waker::wake`, so it can be outside of asyncio thread, that's why `asyncio.get_running_loop` cannot be called here, it must be called in coroutine context (that's the case for `LoopAndFuture::new`).

Storing the loop is not necessary, it can be retrieved by `asyncio.Future.get_loop`, it's just a shortcut to avoid this method call, as we already have the loop object as it's used to create the future. 
I let you decide whether we keep it or call `get_loop`, it's the classical performance vs memory dilemma, and I've no opinion about it.

Reply:
Ah sorry, I just misunderstood the code: You cache the method handle for `get_running_loop`, not its return value. Sorry for the noise.

Reply:
(I also think caching the loop object itself is fine.)